nvim/api.lua

```diff
@@ -0,0 +1,536 @@
+  return string.format('Expected %s[%s] but we found `%s`(%s)', typ, tostring(typ_of), tostring(value), type(value))
+end
+
+--- Implemeted Nvim Basic types
+local Nil = {
+  type_name = 'Nil',
+  basic_type = true,
+  new = function (v) 
+    if type(v) == 'nil' then return nil end
+    return nil, error_convert('Nil', v)
+  end
+}
+
+local Void = {
+  type_name = 'void',
+  basic_type = true,
+  new = function (v) return nil end
+}
+
+local Boolean = {
+  type_name = 'Boolean',
+  basic_type = true,
+  new = function (v)
+    if type(v) == 'boolean' then
+      return v
+    end
+    return nil, error_convert('Boolean', v)
+  end
+}
+
+local Integer = {
+  type_name = 'Integer',
+  basic_type = true,
+  new = function (v)
+    if type(v) == 'number' then
+      return v
+    end
+    return nil, error_convert('Integer', v)
+  end
+}
+
+local Float = {
+  type_name = 'Float',
+  basic_type = true,
+  new = function (v) 
+    if type(v) == 'number' then
+      return v
+    end
+    return nil, error_convert('Float', v)
+  end
+}
+
+local String = {
+  type_name = 'String',
+  basic_type = true,
+  new = function (v) 
+    if type(v) ~= 'string' then
+      return nil, error_convert('String', v)
+    else
+      return tostring(v)
+    end
+  end
+}
+
+local Array = {
+  type_name = 'Array',
+  basic_type = true,
+  new = function (array, of_type) 
+    if type(array) ~= "table" then 
+      return nil, error_convert('Array', array)
+    end
+    if of_type then
+      local array_of = {}
+      for i, value in ipairs(array) do
+        local typed_value, err = of_type.new(value)
+        if not typed_value then 
+          return nil, error_convert_generic('ArrayOf', of_type, value)
+        end
+        array_of[i] = typed_value
+      end
+      return array_of
+    else
+      return util.table_icopy(array)
+    end
+  end
+}
+
+local Dictionary = {
+  type_name = 'Dictionary',
+  basic_type = true,
+  new = function (dictionary) 
+    if type(dictionary) ~= 'table' then
+      return nil, error_convert('Dictionary', dictionary)
+    end
+    return util.table_kvcopy(dictionary)
+  end
+}
+
+local Object = {
+  type_name = 'Object',
+  basic_type = true,
+  new = function(v) return v end,
+}
+
+--- Table to store the basic types in NVim
+local type_tables = {
+  Nil = Nil,
+  Boolean = Boolean,
+  Integer = Integer,
+  Float = Float,
+  String = String,
+  Array = Array,
+  ArrayOf = util.table_kvcopy(Array, {
+    type_generic = 'Array',
+    regex = '(ArrayOf)%((%w+).*%)' -- TODO: We ignore the # of items in the type 
+  }),
+  void = Void, 
+  Dictionary = Dictionary,
+  Object = Object
+}
+
+--- Implemented Ext NVim types
+-- Unknown ext types will still be registered in the Api but
+-- without a fancy wrapper around them.
+local ext_types_table = {
+  Buffer = Buffer,
+  Window = Window,
+  Tabpage = Tabpage
+}
+
+local err_types_table = {
+}
+
+--- Type Buffer
+Buffer.__index = Buffer
```

Comment:
Why are these concrete types needed? What does it gain over plain old lua tables? Is it going to add friction for some applications?